### PR TITLE
Support Ring and Sector ID option in TriPinHexAssemblyGenerator (Reactor Module)

### LINKED ISSUE #21908: Support Ring and Sector ID option in TriPinHexAssemblyGenerator (Reactor Module)
## Reason
 

In PR (#21421), the ring and sector reporting ID (extra integer ID) capability was added to the `PolygonConcentricCircleMeshGenerator`  object in the `reactor` module. But, it did not support the `TriPinHexAssemblyGenerator` in that PR. Recently, this capability was also added to the `TriPinHexAssemblyGenerator`, which allows to manipulate the sub-pin level details in the material arrangement, depletion and post-processing.

## Design

The code sections for  ring and sector ID assignments implemneted in `PolygonConcentricCircleMeshGenerator` were made as functions and moved to the commom base class of both generators, `PolygonMeshGeneratorBase`.
Then, these functions were called in `PolygonConcentricCircleMeshGenerator` and T1riPinHexAssemblyGenerator` to set up the reporting IDs.

## Impact
 
The ring and sector ID generation capability is available in  `PolygonConcentricCircleMeshGenerator` and T1riPinHexAssemblyGenerator` both.

### PR DESCRIPTION
**Pull request by yjung-anl**
closes #21908

### CODE REVIEW

File: modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h [hunk 1]
```diff
@@ -665,4 +665,36 @@ class PolygonMeshGeneratorBase : public MeshGenerator
       const unsigned int intervals,
       const singleBdryLayerParams inner_boundary_layer_params = {0.0, 0.0, 0, 1.0},
       const singleBdryLayerParams outer_boundary_layer_params = {0.0, 0.0, 0, 1.0}) const;
+
+  /**
+   * assign sector extra ids to polygon mesh
+   * @param mesh input mesh where sector extra ids are assined
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:671]:
```suggestion
   * @param mesh input mesh where sector extra ids are assigned
```

File: modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h [hunk 2]
```diff
@@ -665,4 +665,36 @@ class PolygonMeshGeneratorBase : public MeshGenerator
...
+   * @param id_name sector extra ID name
+   * @param num_side number of polygon side
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:673]:
```suggestion
   * @param num_side number of polygon sides
```

File: modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h [hunk 3]
```diff
@@ -665,4 +665,36 @@ class PolygonMeshGeneratorBase : public MeshGenerator
...
+   * @param num_sectors_per_side sector number of each polygon side
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:674]:
```suggestion
   * @param num_sectors_per_side number of sections of each side of the polygon
```

File: modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h [hunk 4]
```diff
@@ -665,4 +665,36 @@ class PolygonMeshGeneratorBase : public MeshGenerator
...
+   */
+  void setSectorExtraIDs(MeshBase & mesh,
+                         const std::string id_name,
+                         const unsigned int num_sides,
+                         const std::vector<unsigned int> num_sectors_per_side);
+
+  /**
+   * assign ring extra ids to polygon mesh
+   * @param mesh  input mesh where ring extra ids are assined
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:683]:
```suggestion
   * @param mesh input mesh where ring extra ids are assigned
```

File: modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h [hunk 5]
```diff
@@ -665,4 +665,36 @@ class PolygonMeshGeneratorBase : public MeshGenerator
...
+   * @param id_name ring extra id name
+   * @param num_sides number of polygon side
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:685]:
```suggestion
   * @param num_sides number of polygon sides
```

File: modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h [hunk 6]
```diff
@@ -665,4 +665,36 @@ class PolygonMeshGeneratorBase : public MeshGenerator
...
+   * @param num_sectors_per_side sector number of each polygon side
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:686]:
```suggestion
   * @param num_sectors_per_side number of sectors of each side of the polygon
```

File: modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C [hunk 7]
```diff
@@ -1408,3 +1408,92 @@ PolygonMeshGeneratorBase::biasTermsCalculator(
...
+                                          const std::string id_name,
+                                          const unsigned int num_sides,
+                                          const std::vector<unsigned int> num_sectors_per_side,
+                                          const std::vector<Real> ring_radii,
+                                          const std::vector<unsigned int> ring_intervals,
+                                          const bool ring_wise_id,
+                                          const bool quad_center_elements)
+{
+  const auto extra_id_index = mesh.add_elem_integer(id_name);
+  auto elem_it = mesh.elements_begin();
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1448]: add a comment that elements are ordered by rings (inner) then by sectors (outer ordering), which is the condition for working with this iterator

File: modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C [hunk 8]
```diff
@@ -1408,3 +1408,92 @@ PolygonMeshGeneratorBase::biasTermsCalculator(
...
+          for (unsigned i = 0; i < nelem_annular_ring; ++i, ++elem_it)
+            (*elem_it)->set_extra_integer(extra_id_index, ir + 1);
+          // update ring id
+          ++ir;
+          // update number of elements in background region of current side.
+          nelem -= nelem_annular_ring;
+        }
+      }
+    }
+    // assign ring id to background region
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:1495]:
```suggestion
    // assign ring id of 0 to the background region
```

File: modules/reactor/src/meshgenerators/TriPinHexAssemblyGenerator.C [hunk 9]
```diff
@@ -68,6 +68,15 @@ TriPinHexAssemblyGenerator::validParams()
...
       "Values of extra integer ID to be assigned to each of the three pin domains.");
+  params.addParam<std::string>("sector_id_name",
+                               "Name of integer (reporting) ID for sector regions to use the "
+                               "reporting ID for azimuthal sector regions of ring geometry block.");
+  params.addParam<std::string>("ring_id_name",
+                               "Name of integer (reporting) ID for ring regions to use the "
+                               "reporting ID for annular regions of ring geometry block.");
+  MooseEnum ring_id_option("block_wise ring_wise", "block_wise");
+  params.addParam<MooseEnum>(
+      "ring_id_assign_type", ring_id_option, "Type of ring ID assignment: block_wise or ring_wise");
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/TriPinHexAssemblyGenerator.C:79]: these parameters could be added by a routine defined in a base class to be shared with the other MG

also add a parameter group for ring/sector IDs

**Review by GiudGiud — CHANGES_REQUESTED** [review]

**Comment by yjung-anl** [comment]
@GiudGiud Thank you for your review. I will address your comments very soon

**Comment by GiudGiud** [comment]
Sounds good. Sorry for the delay in reviewing.

File: modules/reactor/src/meshgenerators/TriPinHexAssemblyGenerator.C (hunk 9 above)
Comment (yjung-anl) [reply modules/reactor/src/meshgenerators/TriPinHexAssemblyGenerator.C]: In the base class, I added a routine to append input parameters for ring / sector IDs as :
```c++
void
PolygonMeshGeneratorBase::addRingAndSectorIDParams(InputParameters & params)
{
  params.addParam<std::string>("sector_id_name",
                               "Name of integer (reporting) ID for sector regions to use the "
                               "reporting ID for azimuthal sector regions of ring geometry block.");
  params.addParam<std::string>("ring_id_name",
                               "Name of integer (reporting) ID for ring regions to use the "
                               "reporting ID for annular regions of ring geometry block.");
  MooseEnum ring_id_option("block_wise ring_wise", "block_wise");
  params.addParam<MooseEnum>(
      "ring_id_assign_type", ring_id_option, "Type of ring ID assignment: block_wise or ring_wise");
  params.addParamNamesToGroup("sector_id_name ring_id_name ring_id_assign_type", "Ring/Sector IDs");
}
```

**Comment by yjung-anl** [comment]
Review comments were addressed

File: modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h [hunk 10]
```diff
@@ -665,4 +665,42 @@ class PolygonMeshGeneratorBase : public MeshGenerator
...
+                         const std::vector<unsigned int> num_sectors_per_side);
+
+  /**
+   * assign ring extra ids to polygon mesh
+   * @param mesh  input mesh where ring extra ids are assigned
+   * @param id_name ring extra id name
+   * @param num_sides number of polygon sides
+   * @param num_sectors_per_side number of sectors of each side of the polygon
+   * @param ring_radii radii of concentric circles
+   * @param ring_intervals number of rings in each circle
```
[the remark this answers is not in the document]
Comment (yjung-anl) [reply modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:694]: Yes. that's correct. I removed the `ring_radii`

**Comment by yjung-anl** [comment]
Comments were addressed

**Review by GiudGiud — APPROVED** [review]
nice. Thanks for the PR!

**Comment by yjung-anl** [comment]
@GiudGiud Thank you for your efforts on this PR!!

**Pull request merged**